internal/daemon/metric/initialize_metrics.go

```diff
@@ -0,0 +1,144 @@
+// Package metric provides functions to initialize the controller specific
+// collectors and hooks to measure metrics and update the relevant collectors.
+package metric
+
+import (
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+)
+
+type LabelNames struct {
+	Service string
+	Method  string
+	Code    string
+}
+
+// ToList is useful for defining label names when creating a new Prometheus vector.
+func (l LabelNames) ToList() []string {
+	return []string{l.Code, l.Service, l.Method}
+}
+
+// ToPromLabels converts the LabelNames struct to a prometheus.Labels object, given the corresponding params.
+func (l LabelNames) ToPromLabels(code string, service string, method string) prometheus.Labels {
+	return prometheus.Labels{l.Code: code, l.Service: service, l.Method: method}
+}
+
+/* The following methods are used to initialize Prometheus histogram vectors for gRPC connections. */
+
+func rangeProtofiles(m map[string][]string, fd protoreflect.FileDescriptor) bool {
+	if fd.Services().Len() == 0 {
+		return true
+	}
+
+	for i := 0; i < fd.Services().Len(); i++ {
+		s := fd.Services().Get(i)
+		if s.Methods().Len() == 0 {
+			continue
+		}
+
+		methods := []string{}
+		for j := 0; j < s.Methods().Len(); j++ {
+			methods = append(methods, string(s.Methods().Get(j).Name()))
+		}
+		m[string(s.FullName())] = methods
+	}
+
+	return true
+}
+
+var allGrpcCodes = []codes.Code{
+	codes.OK, codes.InvalidArgument, codes.PermissionDenied,
+	codes.FailedPrecondition,
+
+	// Codes which can be generated by the gRPC framework
+	codes.Canceled, codes.Unknown, codes.DeadlineExceeded,
+	codes.ResourceExhausted, codes.Unimplemented, codes.Internal,
+	codes.Unavailable, codes.Unauthenticated,
```

Comment:
I'm not clear on this distinction, is it important (the comment)? When we say "the gRPC framework", do we mean the grpc-go libraries? Because these codes can be generated anywhere by anyone.

Reply:
Hmm I don't quite remember the details around this anymore. I think the first four codes may be what we expect to receive from a boundary cluster connection, and the rest are possible codes? I haven't proven this empirically though so I don't endorse my statement

Reply:
I think we can probably just remove the distinction here if we're not confident in it. It would surprise me if we didn't ever return `codes.Unauthenticated` for example